src/verify/header_body.rs

```diff
@@ -274,25 +285,85 @@ pub fn verify(
+    //   contained in the block.
+    // - `Core_execute_block`, which goes through transactions and makes sure that everything is
+    //   valid.
+    //
+    // The first parameter of these two runtime functions is the same: a SCALE-encoded
+    // `(header, body)` where `body` is a `Vec<Extrinsic>`. We perform the encoding ahead of time
+    // in order to re-use it later for the second call.
+    let block_parameter = {
+        // Consensus engines add a seal at the end of the digest logs. This seal is guaranteed to
+        // be the last item. We need to remove it before we can verify the unsealed header.
         let mut unsealed_header = config.block_header.clone();
         let _seal_log = unsealed_header.digest.pop_seal();
 
+        let encoded_body_len = util::encode_scale_compact_usize(config.block_body.len());
+        unsealed_header
+            .scale_encoding()
+            .map(|b| either::Right(either::Left(b)))
+            .chain(iter::once(either::Right(either::Right(encoded_body_len))))
+            .chain(config.block_body.map(either::Left))
+            .fold(Vec::with_capacity(8192), |mut a, b| {
+                // TODO: better capacity ^ ?
+                a.extend_from_slice(AsRef::<[u8]>::as_ref(&b));
+                a
+            })
+    };
+
+    // Start the virtual machine with `BlockBuilder_check_inherents`.
+    let check_inherents_process = {
+        // The second parameter of `BlockBuilder_check_inherents` contains the current timestamp
+        // and current Aura or Babe slot number. The Aura/Babe slot number is no longer used in
+        // more recent runtimes and will lead to errors, but needs to be passed anyway to verify
+        // older blocks.
+        // TODO: remove slot number, see https://github.com/paritytech/smoldot/issues/2096
+        // TODO: uncles?! it's a weird inherent as even in Substrate it's half implemented
+        let inherent_data = inherents::InherentData {
+            timestamp: u64::try_from(config.now_from_unix_epoch.as_millis())
+                .unwrap_or(u64::max_value()),
+            consensus: match config.consensus {
+                ConfigConsensus::AllAuthorized => inherents::InherentDataConsensus::None,
+                // TODO: consider using a function to get current slot from aura/babe module
+                ConfigConsensus::Aura { slot_duration, .. } => {
+                    inherents::InherentDataConsensus::Aura {
+                        slot_number: config.now_from_unix_epoch.as_secs() * 1000
+                            / slot_duration.get(),
+                    }
+                }
+                ConfigConsensus::Babe { .. } => {
+                    inherents::InherentDataConsensus::Babe {
+                        slot_number: u64::max_value(), // TODO: hack /!\ but we don't really care as we're planning to remove this with https://github.com/paritytech/smoldot/issues/2096
```

Comment:
I wanted to be lazy and implement this is hack, because we're going to remove this code with https://github.com/paritytech/smoldot/issues/2096 anyway
Unfortunately, this hack causes the runtime of older blocks to return errors.
I think I'm going to do https://github.com/paritytech/smoldot/issues/2096 first, and then rebase this PR on top.